### ISSUE TITLE
Cannon shell explosion effects sometimes do not appear (but actually they are being spawned at the world origin, i.e. at co-ords zero)

### ISSUE THREAD

**Issue by handsm** [DarklightGames/DarkestHour#653]
Sometimes cannon shell explosion effects do not appear in multi-player. Includes smoke generated by smoke shells, as the smoke emitter gets created as part of the shell's explosion effects.

But in fact the explosion effects are being spawned, it's just that they are produced at the world origin (location co-ords zero). Basnett spotted this and it immediately gave me a vital clue as to what was going wrong.

The problem is specific to cannon shell's as it is caused, indirectly, by an RO system that is unique to cannon shells. Unlike many other projectiles, including bullets, cannon shells  are 'fully replicated', meaning the net client simulated versions of the projectile actor do not get 'torn off' as soon as they are spawned and the server maintains a net channel for the projectile, passing replicated information to it. The server and client versions of the actor do their own local collision detection, triggering a hit and an explosion. But the server version may get to that point first, at which stage it would usually destroy the replicated actor, which would cause it to be destroyed on all clients. If a client actor gets destroyed before running its own collision & explosion, the player wouldn't see any explosion effects; the shell would just wink out of existence.

To counter that, RO added an arbitrary 0.2 second delay before destroying the shell actor. That is designed to give the client time to catch up & do its own explosion. During this time the shell is placed in a kind of suspended animation on the server - although I believe that itself is flawed and may cause problems. As a fallback, when the shell is destroyed on the client, if it hasn't already exploded (bDidExplosionFX hasn't been set to true) then the client spawns explosion effects so the player sees them. Makes sense.

The problem - big one! - is that if the client has to resort to this fallback explosion effects method, it uses its SavedHitLocation and SavedHitNormal as the location and normal for the explosion effects. Basnett noticing the explosions occurring at the world origin was the clue here - means the client has a null value for SavedHitLocation. These two variables are from the RO shell and are recorded with the comment "save the hit info for when the shell is destroyed". But ... they are only set in the Explode() function. And if a net client is falling back to this functionality, by definition the client version of the shell hasn't already exploded, so it can't possibly have set SavedHitLocation!!! A classic catch 22,

**Comment by AndrewTheel** [comment]
Yes! finally strange effects won't be playing at world origin!

**Comment by handsm** [comment]
@AndrewTheel: yeah, you'll love the explanation! Thankfully (for me!) it's an old RO bug. Bonkers! I had a horrible moment when I thought: was it me, could I really have added something so nonsensical? But no.

I've never heard anyone mention that the 'missing' effects were spawning at the world origin. I just thought that sometimes they didn't play at all.

Just testing a few tweak options on the fix, then I'll commit later.

**Comment by handsm** [comment]
Fixed in https://github.com/DarklightGames/DarkestHour/commit/8ef6ac365091f47486c9156a2c1900cf95371c4d.

Deleted the pointless (& dangerous) RO SavedHitLocation and SavedHitNormal variables that were causing the problem. If a net client's shell gets destroyed and made to explode by the server before it has chance to collide and do its own explosion, it now gets its own HitLocation and HitNormal for explosion effects.

Instead of just using the projectile's current location and an arbitrary 'straight up' hit normal, I've used a fairly short forward trace, to try and work out where the projectile was just about to hit. It traces up to 10m ahead, which I think ought to be enough, but if that fails then it falls back to using the current location & a straight up hit normal. I suspect in practice the trace will hit something quite a bit closer than 10m. Using a trace, if it is successful, avoids airburst premature explosion effects and gives us a real hit normal that is useful if spawning the explosion effects properly, e.g. it gives a proper explosion decal.
